lib/task.js

```diff
@@ -404,7 +434,16 @@ function factory(
 
     Task.create = function create(definition, taskOverrides, context) {
         var task = new Task(definition, taskOverrides, context);
-        return task;
+        // If compileOnly is falsey, do rendering in the `.run()` step for
+        // compatibility with how the task runner code is written.
+        if (taskOverrides.compileOnly) {
+            return task.renderAll(task.nodeId, task.definition.options)
+            .then(function() {
```

Comment:
That is a simpler approach, didn't know about `.return`! Though at the same time I'd be worried about causing subtle bugs were other parts of this code ever changed to not pass `self` by reference, but rather another object that requires state changes within the async call chain.